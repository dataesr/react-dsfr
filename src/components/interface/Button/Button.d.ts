import * as React from 'react';

export type ButtonIconPosition = "left" | "right";

export type ButtonClassName = string | Object | any[];

export type ButtonSize = "sm" | "md" | "lg";

export type ButtonChildren = string | React.ReactNode;

export interface ButtonProps {
    secondary?: boolean;
    tertiary?: boolean;
    hasBorder?: boolean;
    icon?: string;
    disabled?: boolean;
    onClick?: (...args: any[])=>any;
    iconPosition?: ButtonIconPosition;
    title?: string;
    styleAsLink?: boolean;
    className?: ButtonClassName;
    size?: ButtonSize;
    children?: ButtonChildren;
    submit?: boolean;
    /**
     * color[0] is background, color[1] is color
     */
    colors?: string[];
}

declare const Button: React.FC<ButtonProps>;

export default Button;

