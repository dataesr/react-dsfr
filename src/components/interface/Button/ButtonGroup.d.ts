import * as React from 'react';

export type ButtonGroupIsInlineFrom = "xs" | "sm" | "md" | "lg";

export type ButtonGroupClassName = string | Object | any[];

export type ButtonGroupSize = "sm" | "md" | "lg";

export type ButtonGroupAlign = "left" | "center" | "right";

export type ButtonGroupChildren = React.ReactNode[] | React.ReactNode;

export interface ButtonGroupProps {
    isInlineFrom?: ButtonGroupIsInlineFrom;
    className?: ButtonGroupClassName;
    size?: ButtonGroupSize;
    align?: ButtonGroupAlign;
    children?: ButtonGroupChildren;
    isReversed?: boolean;
    isEquisized?: boolean;
}

declare const ButtonGroup: React.FC<ButtonGroupProps>;

export default ButtonGroup;

