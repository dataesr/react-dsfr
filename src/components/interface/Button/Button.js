import React, { forwardRef, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import dataAttributes from '../../../utils/data-attributes';
import Icon from '../../foundation/icon/index';

/*
* DSFR v1.0.0
*/
import '@gouvfr/dsfr/dist/css/core.min.css';
import '@gouvfr/dsfr/dist/css/content.min.css';
import '@gouvfr/dsfr/dist/css/forms.min.css';
import '@gouvfr/dsfr/dist/css/inputs.min.css';

/*
* react-ds-fr
*/
import '../../../style/custom.css';

import '@gouvfr/dsfr/dist/css/schemes.min.css';
import '@gouvfr/dsfr/dist/css/buttons.min.css';

const iconSize = {
  sm: 'lg',
  md: 'lg',
  lg: '2x',
};

/**
 * @visibleName Button
 */
const Button = forwardRef((props, ref) => {
  const {
    size,
    secondary,
    disabled,
    title,
    icon,
    iconPosition,
    onClick,
    styleAsLink,
    children,
    className,
    submit,
    colors,
    ...remainingProps
  } = props;

  const _className = classNames(
    `fr-btn--${size}`,
    className,
    {
      'fr-btn': !styleAsLink,
      'fr-link': styleAsLink,
      'fr-fi-icon': !children && icon,
      'fr-btn--secondary': secondary,
    },
  );
  const buttonRef = useRef();
  const myRef = ref || buttonRef;

  useEffect(() => {
    const { current } = myRef;
    const backgroundColor = colors[0];
    const color = colors[1];

    if (current) {
      if (backgroundColor) {
        current.style.backgroundColor = secondary ? color : backgroundColor;
        if (secondary) {
          current.style.boxShadow = `inset 0 0 0 1px ${backgroundColor}`;
        }
      }
      if (color) {
        current.style.color = secondary ? backgroundColor : color;
      }
    }
  }, [colors, myRef, secondary]);

  const _button = (
    <button
      ref={myRef}
      type={submit ? 'submit' : 'button'}
      onClick={onClick}
      className={_className}
      title={title || undefined}
      disabled={disabled}
      {...dataAttributes.getAll(remainingProps)}
    >
      {children}
    </button>
  );
  return icon ? (
    <Icon
      name={icon}
      size={iconSize[size]}
      iconPosition={(children && `${iconPosition}`) || 'center'}
    >
      {_button}
    </Icon>
  ) : _button;
});

Button.defaultProps = {
  size: 'md',
  secondary: false,
  disabled: false,
  iconPosition: 'left',
  icon: '',
  onClick: () => { },
  children: '',
  className: '',
  styleAsLink: false,
  title: null,
  submit: false,
  colors: [],
};

Button.propTypes = {
  secondary: PropTypes.bool,
  icon: PropTypes.string,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  iconPosition: PropTypes.oneOf(['left', 'right']),
  title: PropTypes.string,
  styleAsLink: PropTypes.bool,
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  submit: PropTypes.bool,
  /**
   * color[0] is background, color[1] is color
   */
  colors: PropTypes.arrayOf(PropTypes.string),
};

export default Button;
