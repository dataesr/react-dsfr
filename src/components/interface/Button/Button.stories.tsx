import { ComponentStory } from '@storybook/react';

import Button from './Button';

// https://www.systeme-de-design.gouv.fr/elements-d-interface/composants/bouton

export default {
  title: 'Composants/Bouton - Button',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: 'Le bouton est un élément d\'interaction avec l\'interface permettant à l\'utilisateur d\'effectuer une action.',
      },
    },
  },
};

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  children: 'Button',
};
Primary.storyName = 'Primary';

export const Secondary = Template.bind({});
Secondary.args = {
  children: 'Button',
  secondary: true,
};

export const Tertiary = Template.bind({});
Tertiary.args = {
  children: 'Button',
  tertiary: true,
};

export const WithIcon = Template.bind({});
WithIcon.args = {
  children: 'Button',
  icon: 'ri-alert-fill',
};

export const Disabled = Template.bind({});
Disabled.args = {
  children: 'Button',
  disabled: true,
};

export const NoBorder = Template.bind({});
NoBorder.args = {
  children: 'Button',
  hasBorder: false,
};

export const CustomColor = Template.bind({});
CustomColor.args = {
  children: 'Button',
  colors: ['#f55', '#fff'],
  size: 'lg',
  secondary: true,
  title: 'title',
};
