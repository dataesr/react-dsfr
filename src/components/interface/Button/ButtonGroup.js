import React, { Children } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import dataAttributes from '../../../utils/data-attributes';

/**
 *
 * @visibleName ButtonGroup
 */
const ButtonGroup = ({
  children,
  className,
  size,
  isInlineFrom,
  isReversed,
  isEquisized,
  align,
  ...remainingProps
}) => {
  const content = Children.toArray(children)
    // eslint-disable-next-line react/no-array-index-key
    .map((child, index) => <li key={index}>{child}</li>);
  const _className = classNames('fr-btns-group', {
    [`fr-btns-group--${size}`]: (size !== 'md'),
    [`fr-btns-group--${align}`]: (align !== 'left'),
    'fr-btns-group--inline': (isInlineFrom === 'xs'),
    [`fr-btns-group--inline-${isInlineFrom}`]: (isInlineFrom !== 'xs'),
    'fr-btns-group--inline-reverse': isReversed,
    'fr-btns-group--equisized': isEquisized,
  }, className);
  return (
    <ul
      className={_className}
      {...dataAttributes.getAll(remainingProps)}
    >
      {content}
    </ul>
  );
};

ButtonGroup.defaultProps = {
  size: 'md',
  isInlineFrom: null,
  children: '',
  className: '',
  isReversed: false,
  isEquisized: false,
  align: 'left',
};

ButtonGroup.propTypes = {
  isInlineFrom: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.array,
  ]),
  size: PropTypes.oneOf(['sm', 'md', 'lg']),
  align: PropTypes.oneOf(['left', 'center', 'right']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  isReversed: PropTypes.bool,
  isEquisized: PropTypes.bool,
};

export default ButtonGroup;
