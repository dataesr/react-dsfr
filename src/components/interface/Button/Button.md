```jsx
<Button
    size="sm"
    title="title">basic sm button
</Button>
```

```jsx
<Button
    title="title">basic md button
</Button>
```

```jsx
<Button
    size="lg"
    title="title">basic lg button
</Button>
```

```jsx
<Button
    size="sm"
    secondary
    title="title">secondary sm button
</Button>
```

```jsx
<Button
    disabled
    title="title">disabled button
</Button>
```

```jsx
<Button
    icon="ri-lock-line"
    title="title">icon button
</Button>
```

```jsx
<Button
    icon="ri-alert-fill"
    iconPosition='right'
    size="md"
    title="title">icon on the right
</Button>
```

```jsx
<Button
    icon="ri-alert-fill"
    title="title" />
```

```jsx
import ButtonGroup from './ButtonGroup';

<ButtonGroup
    size="sm"
    isEquisized
    align="center"
    isInlineFrom="md"
>
    <Button
        submit
        data-test-id="save-profile-button"
        icon="ri-check-line"
    >
        Valider
    </Button>
    <Button
        secondary
        icon="ri-close-line"
    >
        Annuler
    </Button>
</ButtonGroup>
```


```jsx
import ButtonGroup from './ButtonGroup';

<ButtonGroup
    size="lg">
  <Button
      secondary
      title="title">secondary sm button
  </Button>
  <Button
      title="title">sm button
  </Button>
</ButtonGroup>
```

