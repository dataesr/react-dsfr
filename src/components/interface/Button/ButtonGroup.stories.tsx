import { ComponentStory } from '@storybook/react';

import ButtonGroup from './ButtonGroup';
import Button from './Button';

export default {
  title: 'Composants/Groupe de Boutons - ButtonGroup',
  component: Button,
  parameters: {
    docs: {
      description: {
        component: 'Le bouton est un élément d\'interaction avec l\'interface permettant à l\'utilisateur d\'effectuer une action. Les boutons dans le contexte d\'un groupe suivent les mêmes règles que le composant bouton',
      },
    },
  },
};

const Template: ComponentStory<typeof ButtonGroup> = (args) => <ButtonGroup {...args} />;

export const VerticalButtons = Template.bind({});
VerticalButtons.args = {
  children: [
    <Button secondary="true">Bouton 1</Button>,
    <Button>Bouton 2</Button>,
    <Button>Bouton 3</Button>,
  ],
};
VerticalButtons.storyName = 'Groupe de boutons vertical';

export const HorizontalButtons = Template.bind({});
HorizontalButtons.args = {
  className: 'fr-btns-group fr-btns-group--inline',
  children: [
    <Button icon="ri-alert-fill">Bouton 1</Button>,
    <Button>Bouton 2</Button>,
    <Button>Bouton 3</Button>,
  ],
};
HorizontalButtons.storyName = 'Groupe de boutons horizontal';
