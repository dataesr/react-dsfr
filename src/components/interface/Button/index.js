import Button from './Button';
import ButtonGroup from './ButtonGroup';

export { Button, ButtonGroup };
