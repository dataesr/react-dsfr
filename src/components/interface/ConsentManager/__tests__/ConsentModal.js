import React from 'react';
import { render, screen } from '@testing-library/react';

import {
  Modal, ModalClose, ModalContent, ModalTitle,
} from '../../Modal';
import { Button } from '../../Button';

describe('<ConsentModal />', () => {
  it('should render ConsentModal properly', () => {
    render(
      <Modal
        data-testid="modal"
        isOpen
        hide={() => jest.fn()}
        id="fr-consent-modal"
        aria-labelledby="fr-consent-modal-title"
      >
        <ModalClose
          hide={jest.fn()}
          title="Close Title"
        >
          Close Label
        </ModalClose>
        <ModalTitle>Title Modal Consent</ModalTitle>
        <ModalContent>
          {null}
          <ul className="fr-consent-manager__buttons fr-btns-group fr-btns-group--right fr-btns-group--inline-sm">
            <li>
              <Button
                title="Button Title"
                onClick={jest.fn()}
              >
                Confirm Label
              </Button>
            </li>
          </ul>
        </ModalContent>
      </Modal>,
    );
    const modal = screen.getByTestId('modal');
    expect(modal).toMatchSnapshot();
  });
});
