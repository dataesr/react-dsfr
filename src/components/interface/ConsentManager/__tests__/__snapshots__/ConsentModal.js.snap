// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConsentModal /> should render ConsentModal properly 1`] = `
<dialog
  aria-labelledby="fr-modal-title-modal"
  class="fr-modal fr-modal--opened"
  data-testid="modal"
  id="fr-consent-modal"
>
  <div
    class="fr-container fr-container--fluid fr-container-md"
  >
    <div
      class="fr-grid-row fr-grid-row--center closing-overlay"
    >
      <div
        class="fr-col-12 fr-col-md-6"
      >
        <div
          class="fr-modal__body"
        >
          <div
            class="fr-modal__header"
          >
            <button
              aria-controls="fr-modal"
              class="fr-link--close fr-link"
              title="Close Title"
              type="button"
            >
              Close Label
            </button>
          </div>
          <div
            class="fr-modal__content"
          >
            <h1
              class="fr-modal__title"
              id="fr-modal-title-modal"
            >
              Title Modal Consent
            </h1>
            <div
              class=""
            >
              <ul
                class="fr-consent-manager__buttons fr-btns-group fr-btns-group--right fr-btns-group--inline-sm"
              >
                <li>
                  <button
                    class="fr-btn--md fr-btn"
                    title="Button Title"
                    type="button"
                  >
                    Confirm Label
                  </button>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</dialog>
`;
